### ISSUE TITLE
VCGCamera CenterPx issues

### ISSUE THREAD

**Issue by eyalshafran** [cnr-isti-vclab/meshlab#84]
I'm using raster layers and the "parameterization + texturing from registered rasters" filter. If I set the CenterPx values to be a none integer in the .mlp file the texture result is always black. This behavior disappears if I round the CenterPx to the nearest integer. I do my own calibration and the center pixels are never an exact integer. This matters if there is camera distortion.

Also, I find the behavior of the CenterPx peculiar. If I have a depth image and color image from the same sensor and I create a point cloud with center pixel x = 300 while the image size is 1200 in the x direction I have to set the CenterPx to 900 (I would have expected 300) to get a result that makes since and the "parameterization + texturing from registered rasters" filter ends up cropping a big part of the image. I don't understand this behavior. Is this a bug or am I missing something?

**Comment by mcallieri** [comment]
confirmed.
The parameter was casted to INT, even though the rest of the code used it as a float.
I corrected the importer and exporter (already committed), and I am checking for other similar problems. 
It will be available in the next release.

As for the second point, BEWARE: our camera and texture projection code assumes the center pixel is ALMOST in the middle of the image. non-central perspective images do require much more care, and additional checks. So, we cannot guarantee the behavior of the tool in those cases.

**Comment by eyalshafran** [comment]
Thanks.

Regarding the second point - the large offset was just a test to understand the behavior. I create the point cloud from depth data and the camera intrinsics. I couldn't get good color overlay so I decided to modify some parameters systematically to find the problem.  I was surprised to find out that Center pixels parameters that I use for point cloud creation are not the actual values that need to be inserted into the .mlp file. I'm still not sure I understand the logic. If I create a point cloud with xc=300 why do I need to set the CenterPx to 900 (1200 is the image size) for good color overlay? Is there more detailed documentation regarding the VCGCamera parameters?

**Comment by eyalshafran** [comment]
It is also unclear how to set the distortion coefficients. Is it normalized or unnormalized units? Something seems to be going wrong here. I simulated a point cloud with distortion that also has color per vertex. I use the color image per vertex as a raster layer. If I simulate a point cloud with no distortion the color overlay of the raster layer is perfect (I still have to use CenterPx = image size - actual CenterPx a stated above). If I use the distortion coefficients in the simulation then there is a mismatch with the overlay the further I get from the center suggesting that the distortion correction is not working properly. I verified mathematically that when I project back points from the point cloud to the camera and use the distortion correction I get the correct pixel on the camera. I tried both normalized and unnormalized distortion coefficients. The normalized units seem to be way too large so I'm assuming that is not correct. The unnormalized units are so small it has no influence.

I think there is a bug in the projection to the camera. 

Thank you!

**Comment by mcallieri** [comment]
The camera model used in MeshLab is extremely simple, and probably also extremely limited. The camera was modeled after the data structures used by TSAI. It is working in this way to cope with basic image projection and camera calibration/orientation, MeshLab is not a photogrammetry tool, and we know there are some buggy/limitations: distorsion is not always used (our fault), and non-central perspectives are probably incorrectly managed. As we are able to import and aling photos, texture-map and project color, import from bundler and other similar software, it is enough for our use.
I really cannot understand in your messages what is the problem about the center pixel... if you have a 1200 pixel image with optical center at 300, it is a non-central perspective, which probably will not work in MeshLab.

**Comment by eyalshafran** [comment]
Thanks for the clarification. 

I thought that the whole point of the raster layer is photogrammetry. I have been using Meshlab as a photogrammetry tool and it actually works great under certain conditions. It is good to know about the distortion - I now undistort my image using Matlab and use that as a raster layer (and I set the distortion to 0 in the .mlp file). Things are working well now. The "parameterization + texturing from registered rasters" filter is especially useful in my case.

The center pixel at 300 was just an extreme example in order to understand how to format the .mlp file properly due to lack of documentation. This is not something that I use. In practice, my center pixel is shifted from the image center by 0-20 pixels. I still don't understand why I need to set the center pixel = image size - center pixel from calibration (instead of just center pixel = center pixel from calibration) but I guess I got things to work so it doesn't really matter. I'm sure I'm not the last to have this issue so a short tutorial on how to set the VCGCamera properties from a known set of calibration parameters would be useful.
